### PR TITLE
lgc: PatchReadFirstLane cleanups and one bugfix

### PR DESCRIPTION
**Pull request by nhaehnle**
Please see individual commit messages

Fixes: https://github.com/GPUOpen-Drivers/llpc/issues/2746

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1d71a7a296c9a99c75bdbdc10eaa3f45d1efc150
<details><summary>CTS tests (Failed: 0/139021)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69517 (59.2%)</li>
<li>Failed: 0/69517 (0.0%)</li>
<li>Not Supported: 28373/69517 (40.8%)</li><li>Warnings: 0/69517 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69504 (59.2%)</li>
<li>Failed: 0/69504 (0.0%)</li>
<li>Not Supported: 28326/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7c6614d03c09f577e1e84e4b1f3bd2a6bf656507
<details><summary>CTS tests (Failed: 0/139021)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69517 (59.2%)</li>
<li>Failed: 0/69517 (0.0%)</li>
<li>Not Supported: 28373/69517 (40.8%)</li><li>Warnings: 0/69517 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69504 (59.2%)</li>
<li>Failed: 0/69504 (0.0%)</li>
<li>Not Supported: 28326/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchReadFirstLane.cpp [hunk 1]
```diff
@@ -397,16 +396,17 @@ void ReadFirstLaneOptimizer::findBestInsertLocation(const SmallVectorImpl<Instru
...
+      assert(mapIt != m_uniformDivergentUsesMap.end());
       const auto &divergentOperands = mapIt->second;
       if (divergentOperands.empty())
         break; // no further propagation possible
 
       if (divergentOperands.size() == 1) {
         // There is only a single operand, we can jump to it directly.
         current = divergentOperands[0];
+
+        if (!m_uniformDivergentUsesMap.count(current))
```
Comment (ruiling) [inline lgc/patch/PatchReadFirstLane.cpp:408]: I think you need to put this after the if/else to cover both cases.

Comment (nhaehnle) [reply lgc/patch/PatchReadFirstLane.cpp:408]: The else case has an inner loop with the invariant that `queue.size() >= 2` at the top of the loop; and queue members are only popped if they pass the m_uniformDivergentUsesMap check. Okay, but on second thought, `queue[0]` may end up going unchecked. So I think you're right.

**Comment by amdvlk-admin** [comment]
### Test summary for commit ef766afc001d0adf083ea0eefb62797a726eea1e
<details><summary>CTS tests (Failed: 0/139021)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 41144/69517 (59.2%)</li>
<li>Failed: 0/69517 (0.0%)</li>
<li>Not Supported: 28373/69517 (40.8%)</li><li>Warnings: 0/69517 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 41178/69504 (59.2%)</li>
<li>Failed: 0/69504 (0.0%)</li>
<li>Not Supported: 28326/69504 (40.8%)</li><li>Warnings: 0/69504 (0.0%)</li>
</ul></details>


</ul></details>

**Review by ruiling — APPROVED** [review]
LGTM

**Review by dstutt — APPROVED** [review]

**Pull request merged**